src/Microsoft.AspNetCore.SpaServices.Extensions/Npm/EventedStreamReader.cs

```diff
@@ -0,0 +1,134 @@
+
+namespace Microsoft.AspNetCore.NodeServices.Util
+{
+    /// <summary>
+    /// Wraps a <see cref="StreamReader"/> to expose an evented API, issuing notifications
+    /// when the stream emits partial lines, completed lines, or finally closes.
+    /// </summary>
+    internal class EventedStreamReader
+    {
+        public delegate void OnReceivedChunkHandler(ArraySegment<char> chunk);
+        public delegate void OnReceivedLineHandler(string line);
+        public delegate void OnStreamClosedHandler();
+
+        public event OnReceivedChunkHandler OnReceivedChunk;
+        public event OnReceivedLineHandler OnReceivedLine;
+        public event OnStreamClosedHandler OnStreamClosed;
+
+        private readonly StreamReader _streamReader;
+        private readonly StringBuilder _linesBuffer;
+
+        public EventedStreamReader(StreamReader streamReader)
+        {
+            _streamReader = streamReader ?? throw new ArgumentNullException(nameof(streamReader));
+            _linesBuffer = new StringBuilder();
+            Task.Factory.StartNew(Run);
+        }
+
+        public Task<Match> WaitForMatch(Regex regex, int timeoutMilliseconds = 0)
+        {
+            var tcs = new TaskCompletionSource<Match>();
+            var completionLock = new object();
+
+            OnReceivedLineHandler onReceivedLineHandler = null;
+            OnStreamClosedHandler onStreamClosedHandler = null;
+
+            void ResolveIfStillPending(Action applyResolution)
+            {
+                lock (completionLock)
+                {
+                    if (!tcs.Task.IsCompleted)
+                    {
+                        OnReceivedLine -= onReceivedLineHandler;
+                        OnStreamClosed -= onStreamClosedHandler;
+                        applyResolution();
+                    }
+                }
+            }
+
+            onReceivedLineHandler = line =>
+            {
+                var match = regex.Match(line);
+                if (match.Success)
+                {
+                    ResolveIfStillPending(() => tcs.SetResult(match));
+                }
+            };
+
+            onStreamClosedHandler = () =>
+            {
+                ResolveIfStillPending(() => tcs.SetException(new EndOfStreamException()));
+            };
+
+            OnReceivedLine += onReceivedLineHandler;
+            OnStreamClosed += onStreamClosedHandler;
+
+            if (timeoutMilliseconds > 0)
+            {
+                var timeoutToken = new CancellationTokenSource(timeoutMilliseconds);
+                timeoutToken.Token.Register(() =>
+                {
+                    ResolveIfStillPending(() => tcs.SetCanceled());
+                });
+            }
+
+            return tcs.Task;
+        }
+
+        private async Task Run()
+        {
+            var buf = new char[8 * 1024];
+            while (true)
+            {
+                var chunkLength = await _streamReader.ReadAsync(buf, 0, buf.Length);
```

Comment:
Why not use ReadLineAsync?

Reply:
Because we want notifications of both partial-line chunks (for progress updates) and complete lines (that can be forwarded to the `ILogger`).